### PR TITLE
2020 HUD HMIS CSV spec

### PR DESCRIPTION
**Pull request by eanders**
This implements the an importer, exporter, and storage for the HUD 2020 HMIS CSV spec.  Of note are the removal of Geography and the addition of Event, CurrentLivingSituation, User, Assessment, AssessmentQuestion, and AssessmentResult.

The test suite is a bit thin currently, and this does *not* implement any data migration from the 6.12 spec, nor any changes to the UI that will be needed to show the new locations of geography data, but this should be enough for us to accept uploads from other vendors and generate exports in the correct format.

### CODE REVIEW

File: app/models/concerns/export/hmis_twenty_twenty/shared.rb [hunk 1]
```diff
@@ -0,0 +1,313 @@
...
+      row
+    end
+    row
+  end
+
+  # All HUD Keys will need to be replaced with our IDs to make them unique across data sources.
+  def columns_to_pluck
+    @columns_to_pluck = hud_csv_headers.map do |k|
+      case k
+      when self.class.hud_key.to_sym
```
[the remark this answers is not in the document]
Comment (eanders) [reply app/models/concerns/export/hmis_twenty_twenty/shared.rb]: sadly, the convention for case statements in ruby is to not indent the where lines.  I still don't know why, but I'm following it until I have a compelling reason not to.

File: app/models/concerns/import/hmis_twenty_twenty/shared.rb [hunk 2]
```diff
@@ -0,0 +1,292 @@
...
+        pluck(
+          self.hud_key,
+          :DateUpdated,
+          :DateDeleted,
+          :id,
+          :source_hash,
+          :pending_date_deleted,
+          :data_source_id
+        ).map do |key, updated_at, deleted_at, id, source_hash, pending_date_deleted, data_source_id|
+          [key, OpenStruct.new(
```
[the remark this answers is not in the document]
Comment (eanders) [reply app/models/concerns/import/hmis_twenty_twenty/shared.rb]: OpenStruct allows access via dot notation.  Generally it's not a significant speed decrease, but can be a huge win for readability.

File: app/models/grda_warehouse/import/hmis_twenty_twenty/geography.rb [hunk 3]
```diff
@@ -0,0 +1,28 @@
+###
+# Copyright 2016 - 2019 Green River Data Analysis, LLC
+#
+# License detail: https://github.com/greenriver/hmis-warehouse/blob/master/LICENSE.md
+###
+
+module GrdaWarehouse::Import::HmisTwentyTwenty
+  class Geography < GrdaWarehouse::Hud::Geography
```
[the remark this answers is not in the document]
Comment (eanders) [reply app/models/grda_warehouse/import/hmis_twenty_twenty/geography.rb]: Correct, this must have slipped back in.

File: app/models/help/private.rb [hunk 4]
```diff
@@ -0,0 +1,10 @@
+###
+# Copyright 2016 - 2019 Green River Data Analysis, LLC
+#
+# License detail: https://github.com/greenriver/hmis-warehouse/blob/master/LICENSE.md
+###
+
+module Help
```
[the remark this answers is not in the document]
Comment (eanders) [reply app/models/help/private.rb]: nope.  Thanks.

**Comment by eanders** [comment]
In terms of copy paste vs re-factor.  I'm still leaning toward copy-paste.  These importers/exporters tend to last a year or two and then get replaced.  The overlap where we need to support the old version is roughly 6 months.  With copy-paste, it's relatively easy to yank out the old one when it's no longer necessary and not worry if we're damaging the new versions.  Also, when we build out a new one we can begin yanking out anything that is no longer necessary without worrying about breaking the existing versions.

**Comment by awisema** [comment]
There are some naming conflicts in the 2020 vs 6.1.1 tests:
```
WARNING: Shared example group 'project setup' has been previously defined at:
  /Users/sandy/Sites/hmis-warehouse/spec/models/exporters/hmis_six_one_one/project_setup.rb:1
...and you are now defining it at:
  /Users/sandy/Sites/hmis-warehouse/spec/models/exporters/hmis_twenty_twenty/project_setup.rb:1
The new definition will overwrite the original one.
WARNING: Shared example group 'multi-project tests' has been previously defined at:
  /Users/sandy/Sites/hmis-warehouse/spec/models/exporters/hmis_six_one_one/multi_project_tests.rb:1
...and you are now defining it at:
  /Users/sandy/Sites/hmis-warehouse/spec/models/exporters/hmis_twenty_twenty/multi_project_tests.rb:1
The new definition will overwrite the original one.
WARNING: Shared example group 'enrollment setup' has been previously defined at:
  /Users/sandy/Sites/hmis-warehouse/spec/models/exporters/hmis_six_one_one/enrollment_setup.rb:1
...and you are now defining it at:
  /Users/sandy/Sites/hmis-warehouse/spec/models/exporters/hmis_twenty_twenty/enrollment_setup.rb:1
The new definition will overwrite the original one.
WARNING: Shared example group 'multi-enrollment tests' has been previously defined at:
  /Users/sandy/Sites/hmis-warehouse/spec/models/exporters/hmis_six_one_one/multi_enrollment_tests.rb:1
...and you are now defining it at:
  /Users/sandy/Sites/hmis-warehouse/spec/models/exporters/hmis_twenty_twenty/multi_enrollment_tests.rb:1
The new definition will overwrite the original one.
WARNING: Shared example group 'enrollment dates setup' has been previously defined at:
  /Users/sandy/Sites/hmis-warehouse/spec/models/exporters/hmis_six_one_one/enrollment_dates_setup.rb:1
...and you are now defining it at:
  /Users/sandy/Sites/hmis-warehouse/spec/models/exporters/hmis_twenty_twenty/enrollment_dates_setup.rb:1
The new definition will overwrite the original one.
WARNING: Shared example group 'single-project tests' has been previously defined at:
  /Users/sandy/Sites/hmis-warehouse/spec/models/exporters/hmis_six_one_one/single_project_tests.rb:1
...and you are now defining it at:
  /Users/sandy/Sites/hmis-warehouse/spec/models/exporters/hmis_twenty_twenty/single_project_tests.rb:1
The new definition will overwrite the original one.
WARNING: Shared example group 'single-enrollment tests' has been previously defined at:
  /Users/sandy/Sites/hmis-warehouse/spec/models/exporters/hmis_six_one_one/single_enrollment_tests.rb:1
...and you are now defining it at:
  /Users/sandy/Sites/hmis-warehouse/spec/models/exporters/hmis_twenty_twenty/single_enrollment_tests.rb:1
The new definition will overwrite the original one.
WARNING: Shared example group 'project continuum override setup' has been previously defined at:
  /Users/sandy/Sites/hmis-warehouse/spec/models/exporters/hmis_six_one_one/project_continuum_override_setup.rb:1
...and you are now defining it at:
  /Users/sandy/Sites/hmis-warehouse/spec/models/exporters/hmis_twenty_twenty/project_continuum_override_setup.rb:1
The new definition will overwrite the original one.
WARNING: Shared example group 'project continuum override tests' has been previously defined at:
  /Users/sandy/Sites/hmis-warehouse/spec/models/exporters/hmis_six_one_one/project_continuum_override_tests.rb:1
...and you are now defining it at:
  /Users/sandy/Sites/hmis-warehouse/spec/models/exporters/hmis_twenty_twenty/project_continuum_override_tests.rb:1
The new definition will overwrite the original one.
WARNING: Shared example group 'project type override setup' has been previously defined at:
  /Users/sandy/Sites/hmis-warehouse/spec/models/exporters/hmis_six_one_one/project_type_override_setup.rb:1
...and you are now defining it at:
  /Users/sandy/Sites/hmis-warehouse/spec/models/exporters/hmis_twenty_twenty/project_type_override_setup.rb:1
The new definition will overwrite the original one.
WARNING: Shared example group 'project type override tests' has been previously defined at:
  /Users/sandy/Sites/hmis-warehouse/spec/models/exporters/hmis_six_one_one/project_type_override_tests.rb:1
...and you are now defining it at:
  /Users/sandy/Sites/hmis-warehouse/spec/models/exporters/hmis_twenty_twenty/project_type_override_tests.rb:1
The new definition will overwrite the original one.
```

File: app/models/concerns/import/hmis_twenty_twenty/shared.rb [hunk 5]
```diff
@@ -0,0 +1,292 @@
...
+      # the batch import fails to correctly guess the timezone, force these into useful times
+      row[:DateUpdated] = row[:DateUpdated]&.to_time
+      row[:DateCreated] = row[:DateCreated]&.to_time
+      row[:DateDeleted] = row[:DateDeleted]&.to_time
+      row = translate_to_db_headers(row)
+      row
+    end
+
+    def should_add? existing
+      existing.to_h.blank?
```
[the remark this answers is not in the document]
Comment (eanders) [reply app/models/concerns/import/hmis_twenty_twenty/shared.rb]: I can't remember the thinking behind this originally.  Perhaps it forces an evaluation of the CSV row?  I'm leaving it for now, but will add a comment so we can investigate later.  (6.12 version has the same code.)

**Pull request merged**